include/swift/AST/Decl.h

```diff
@@ -7496,6 +7573,16 @@ class SubscriptDecl : public GenericContext, public AbstractStorageDecl {
   /// implies.
   ObjCSubscriptKind getObjCSubscriptKind() const;
 
+  /// Determines, caches, and returns whether the decl can be used to satisfy an
+  /// `@dynamicMemberLookup` requirement (and if so, how).
+  DynamicMemberLookupSubscriptEligibility
+  getDynamicMemberLookupSubscriptEligibility();
```

Comment:
Instead of exposing this enum, I suppose it is also possible to restore `isValidDynamicMemberLookupSubscript()` et. al., but I do think it's nicer to have a type-safe interface that's harder to typo.